### ISSUE TITLE
Signed bit field data types?

### ISSUE THREAD

**Issue by JaapAap** [kaitai-io/kaitai_struct#121]
Are signed bit field data types supported? I need to read a 17 bit wide field (two's complement) into a signed integer but the `bx `fields all map into unsigned integers?

**Comment by GreyCat** [comment]
Not directly. With non-standard integers there are so many variations, so it's probably more practical to define your own using value instance, i.e. something like that:

```yaml
types:
  my_int17:
    seq:
      - id: minus
        type: b1
      - id: raw
        type: b16
    instances:
      value:
        value: minus ? ~raw : raw
```

**Comment by JaapAap** [comment]
Thanks for that (and other replies), much appreciated. I also noted that the bit order in my case is different from the one used by default. Let me take the example from the user guide. I would like to use  a specification similar to the one in the example (which maps well with the documentation of the protocol I am implementing):
```
seq:
  - id: a
    type: b5
  - id: b
    type: b9
    # 3 bits + 6 bits
  - id: c
    type: b2
```

but rather than taking the bits from the standard layout  as being used now:

```
76543210 76543210
aaaaabbb bbbbbbcc
```

I would need to have these bits extracted from a layout in which the bits are in ordered as:

```
76543210 76543210
bbbaaaaa ccbbbbbb
```

Could this be achieved elegantly somehow?

**Comment by GreyCat** [comment]
Not really *elegantly*, but you can just do basic bit math:

```yaml
seq:
  - id: b_part1
    type: b3
  - id: a
    type: b5
  - id: c
    type: b2
  - id: b_part2
    type: b6
instances:
  b:
    value: (b_part1 << 6) + b_part2
```

If your format requires some other composition, may be `value: (b_part2 << 3) + b_part1` will be the correct answer, or even something more exotic. There's no one-size-fits-all solution here, unfortunately. We've started to discuss some better alternative in #58 — feel free to drop a line there if you have some ideas :)

**Comment by JaapAap** [comment]
Thanks - while the math isn't that hard, it sort of defeats the purpose of using a domain specific language - which is why I would like to use kaitai. I'll surely check the other discussion - I understand that it will be difficult to cover all cases (but I believe that the one at hand is typical for many streaming data protocols).

**Comment by JaapAap** [comment]
I decided to implement 3 more bit-field types (for the C++ runtime), and learn a bit more about kaitai in the mean time - covering n-bit two's complement signed integers as well as the two different ways of 'counting' the bytes (resulting in 4 types of fields, tentively indicated by `bnn` (the currently existing type), `sbnn` (the signed version) as well as `lbnn` `slbnn` for the versions in which bits are counted in LE order. I am not sure whether there would be any of interest in incorporating this functionality but would obviously be happy to share it.

**Comment by GreyCat** [comment]
Send in the pull request :) I believe lots of people would be interesting in seeing in — given on how many comment we've got in #9.

**Comment by JaapAap** [comment]
I will give that a shot - but I guess that should be two separate but related pull requests - one for the runtime and one for the compiler - could these be coupled somehow? 

Before doing so, could I ask you whether the following makes sense to you: I currently call `align_to_byte()` whenever the next field is (1) has a different way of counting the bits in a bit-field, or (2) when the next field is byte based (as is the current behaviour). It actually turned out that in some of my data messages that I am encoding, both types are used, but in separate sections, and they can obviously not be mixed safely under all conditions - hence I decided to at least make bit-field sections start on a byte boundary. 

Also, I am not overly happy with the naming convention - but not necessarily have a better alternative. Maybe 'signed (s)' should reflect more clearly that it is two's complement based ('tcbnn' could be an alternative to 'sbnn') and the Little Endian qualifier (l) might also be replaced by something less ambiguous. Opinions are more than welcome.

**Comment by GreyCat** [comment]
> I will give that a shot - but I guess that should be two separate but related pull requests - one for the runtime and one for the compiler - could these be coupled somehow?

Not really, but you can add reference to an URL or commit ID from one to another — that would certainly help :)

> I currently call `align_to_byte()` whenever the next field is (1) has a different way of counting the bits in a bit-field, or (2) when the next field is byte based (as is the current behaviour).

Makes sense, yeah. Actually, it would really help if you'd provide some examples of where it's used in real-life environment as well. Especially if it's clear that it's used at more than a single occurence ;)

> Also, I am not overly happy with the naming convention

Something like `bXXle` / `bXXbe` comes to mind, in lines with existing notation for integers. Not sure about signed stuff as well :(

**Comment by JaapAap** [comment]
I have implemented signed and unsigned bitfields for both LE and BE byte ordering (with the current bitfield implementation essentially equal to unsigned/BE, using ub and sb to denote signedness and trailing be and le for endianess) . The issue I now run into is the following. For BE, the bit order witin a byte is 1-2-3-4-5-6-7-8, while for LE it is 8-7-6-5-4-3-2-1. Hence, the interpretation of the sequence 

```
seq:
  - id: bit1
    type: ub1
  - id: bit2
    type: ub1
```
...

will be different for 

```
meta:
  endian: be
```

and 

```
meta:
  endian: le
```

as it will select different bits. This is all well for the newly introduced types, but for the currently built-in bitfield, the behaviour is always BE/unsigned", even if LE endianness is specified, which can easily become confusing.

I just would like to ask your opinion on dealing with this - I suppose that the current behaviour could be kept but in that case it might have to be better documented?

**Comment by arekbulski** [comment]
In Construct, BitsInteger supports signed integers (2-complement).
https://construct.readthedocs.io/en/latest/api/numerics.html#construct.BitsInteger

**Comment by Barracuda72** [comment]
> In Construct, BitsInteger supports signed integers (2-complement).
> https://construct.readthedocs.io/en/latest/api/numerics.html#construct.BitsInteger

I'm voting for this. 2's complement is one of the most standard way to represent an integer, even in non-standard integer sizes.
